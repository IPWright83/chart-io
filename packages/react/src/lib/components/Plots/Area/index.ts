export * from "./Area";
export * from "./StackedArea";
export * from "./Areas";
