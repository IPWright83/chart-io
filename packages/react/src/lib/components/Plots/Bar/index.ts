export * from "./Bar";
export * from "./StackedBars";
export * from "./GroupedBars";
export * from "./Bars";
