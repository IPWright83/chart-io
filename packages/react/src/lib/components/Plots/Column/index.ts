export * from "./Column";
export * from "./StackedColumn";
export * from "./GroupedColumn";
export * from "./Columns";
