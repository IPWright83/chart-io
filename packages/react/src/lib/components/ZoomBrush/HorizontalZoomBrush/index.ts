import { HorizontalZoomBrush as HorizontalZoomBrushBase, IHorizontalZoomBrushProps } from "./HorizontalZoomBrush";
import { withSVG } from "../../../hoc";

export const HorizontalZoomBrush = withSVG<IHorizontalZoomBrushProps>(HorizontalZoomBrushBase, "horizontal-zoom-brush");

export type { IHorizontalZoomBrushProps };
