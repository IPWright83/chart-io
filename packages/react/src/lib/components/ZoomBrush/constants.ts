export const DEFAULT_BRUSH_MARGIN = { left: 0, top: 10, right: 0, bottom: 10 };
