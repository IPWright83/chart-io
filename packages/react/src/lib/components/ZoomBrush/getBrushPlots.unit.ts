import { getBrushPlots } from "./getBrushPlots";

const PlotWithBrush = { props: {}, type: { isPlot: true } };
const PlotWithoutBrush = { props: {}, type: { isPlot: false } };
const NullComponent = null;

describe("getBrushPlots", () => {
  it("should filter out non-plot components and return plot components", () => {
    const children = [PlotWithBrush, PlotWithoutBrush, NullComponent, PlotWithBrush];
    const result = getBrushPlots(children);
    expect(result).toEqual([PlotWithBrush, PlotWithBrush]);
  });

  it("should return an empty array if there are no plot components", () => {
    const children = [NullComponent, NullComponent, NullComponent];
    const result = getBrushPlots(children);
    expect(result).toEqual([]);
  });

  it("should return an empty array if children is not an array", () => {
    const children = PlotWithoutBrush;
    const result = getBrushPlots(children);
    expect(result).toEqual([]);
  });
});
