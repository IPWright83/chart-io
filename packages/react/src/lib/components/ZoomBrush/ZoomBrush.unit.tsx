import { d3 } from "@chart-io/core";
import React from "react";

import { Bar, Line } from "../Plots";
import { renderChart, wait } from "../../testUtils";

import { ZoomBrush } from "./ZoomBrush";

describe("ZoomBrush", () => {
    it("renders if not required for the given plots", async () => {
        const data = [
            { x: 5, y: 5 },
            { x: 10, y: 10 },
        ];
        const scales = {
            x: d3.scaleLinear().domain([0, 20]).range([0, 100]),
            y: d3.scaleLinear().domain([0, 20]).range([0, 100]),
        };

        const { asFragment } = await renderChart({
            children: (
                <ZoomBrush size={100}>
                    <Line x="x" y="y" />
                </ZoomBrush>
            ),
            data,
            scales,
        });

        // Line plots sometimes take a short while to draw
        await wait(100);

        expect(asFragment()).toMatchSnapshot();
    });

    it("doesn't render if not required for the given plots", async () => {
        const data = [
            { x: 5, y: 5 },
            { x: 10, y: 10 },
        ];
        const scales = {
            x: d3.scaleLinear().domain([0, 20]).range([0, 100]),
            y: d3.scaleLinear().domain([0, 20]).range([0, 100]),
        };

        const { asFragment } = await renderChart({
            children: (
                <ZoomBrush size={100}>
                    <Bar x="x" y="y" />
                </ZoomBrush>
            ),
            data,
            scales,
        });

        expect(asFragment()).toMatchSnapshot();
    });

    it("doesn't render if set to currently un-supported overlay", async () => {
        const data = [
            { x: 5, y: 5 },
            { x: 10, y: 10 },
        ];
        const scales = {
            x: d3.scaleLinear().domain([0, 20]).range([0, 100]),
            y: d3.scaleLinear().domain([0, 20]).range([0, 100]),
        };

        const { asFragment } = await renderChart({
            children: (
                <ZoomBrush size={100} type="overlay">
                    <Line x="x" y="y" />
                </ZoomBrush>
            ),
            data,
            scales,
        });

        // Line plots sometimes take a short while to draw
        await wait(100);

        expect(asFragment()).toMatchSnapshot();
    });
});
