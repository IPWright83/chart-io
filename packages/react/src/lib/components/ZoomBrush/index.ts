import { ZoomBrush as ZoomBrushBase, IZoomBrushProps } from "./ZoomBrush";
import { withSVG } from "../../hoc";

export const ZoomBrush = withSVG<IZoomBrushProps>(ZoomBrushBase, "zoom-brush");

export type { IZoomBrushProps };
