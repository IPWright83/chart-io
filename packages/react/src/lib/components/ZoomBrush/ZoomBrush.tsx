import { IMargin } from "@chart-io/types";
import React from "react";
import { DEFAULT_BRUSH_MARGIN } from "./constants";

import { getBrushPlots } from "./getBrushPlots";
import { HorizontalZoomBrush } from "./HorizontalZoomBrush";

export interface IZoomBrushProps {
    /**
     * The child components for the chart
     */
    children?: JSX.Element | JSX.Element[];
    /**
     * The number of pixels the brush should take. This is the height
     * for a horizontal brush or the width for a vertical brush.
     */
    size?: number;

    /**
     * The type of brush to use. Either an "inline" brush which sits beside the
     * chart or an "overlay" brush which sits on top of the chart
     */
    type: "inline" | "overlay";

    /**
     * The margin to apply around the brush
     * @default { left: 0, top: 10, right: 0, bottom: 10 }
     */
    margin?: IMargin;
}

/**
 * Represents a zoomable Brush component on the chart
 * @return The Brush component
 */
export function ZoomBrush({ children, size = 60, type = "overlay", margin = DEFAULT_BRUSH_MARGIN }: IZoomBrushProps) {
    // Find all the interactive plots
    const plots = getBrushPlots(children);

    const needsXBrush = plots.some((c) => c.type.brush.horizontal);
    // const needsYBrush = plots.some((c) => c.type.brush.vertical);

    if (type === "inline" && needsXBrush) {
        return (
            <HorizontalZoomBrush height={size} margin={margin}>
                {plots}
            </HorizontalZoomBrush>
        );
    }
}
