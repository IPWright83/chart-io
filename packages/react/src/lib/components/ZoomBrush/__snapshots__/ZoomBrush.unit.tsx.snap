// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ZoomBrush doesn't render if not required for the given plots 1`] = `
<DocumentFragment>
  <svg />
</DocumentFragment>
`;

exports[`ZoomBrush doesn't render if set to currently un-supported overlay 1`] = `
<DocumentFragment>
  <svg />
</DocumentFragment>
`;

exports[`ZoomBrush renders if not required for the given plots 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-horizontal-zoom-brush"
    >
      <g
        fill="none"
        pointer-events="all"
        transform="translate(0, 210)"
      >
        <g
          class="g-plot line"
        >
          <path
            class="line"
            d="M25,25L50,50                      "
            style="fill: none; pointer-events: none; stroke: rgb(153, 193, 220); stroke-width: 2px;"
          />
          <path
            class="line"
            d="M0,0L100,0"
            style="fill: none; pointer-events: none;"
          />
        </g>
        <rect
          class="overlay"
          cursor="crosshair"
          height="100"
          pointer-events="all"
          width="200"
          x="0"
          y="0"
        />
        <rect
          class="selection"
          cursor="move"
          fill="#777"
          fill-opacity="0.3"
          shape-rendering="crispEdges"
          stroke="#fff"
          style="display: none;"
        />
        <rect
          class="handle handle--w"
          cursor="ew-resize"
          style="display: none;"
        />
        <rect
          class="handle handle--e"
          cursor="ew-resize"
          style="display: none;"
        />
      </g>
    </g>
  </svg>
</DocumentFragment>
`;
